flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java

```diff
@@ -184,17 +189,19 @@ private SnapshotDirectory prepareLocalSnapshotDirectory(long checkpointId) throw
 			LocalRecoveryDirectoryProvider directoryProvider = localRecoveryConfig.getLocalStateDirectoryProvider();
 			File directory = directoryProvider.subtaskSpecificCheckpointDirectory(checkpointId);
 
-			if (directory.exists()) {
-				FileUtils.deleteDirectory(directory);
-			}
-
-			if (!directory.mkdirs()) {
+			if (!directory.exists() && !directory.mkdirs()) {
 				throw new IOException("Local state base directory for checkpoint " + checkpointId +
 					" already exists: " + directory);
 			}
 
 			// introduces an extra directory because RocksDB wants a non-existing directory for native checkpoints.
-			File rdbSnapshotDir = new File(directory, "rocks_db");
+			// append operatorIdentifier here to solve directory collision problem when two stateful operators chained in one task.
+			String subDir = operatorIdentifier.replaceAll("[^a-zA-Z0-9\\-]", "_");
```

Comment:
I would suggest to move the string sanatizing into a separate method and we could use a precompiled regex that we can reuse every time over `replaceAll`.

Reply:
Thinking about it a bit more, I wonder if it is a good idea to the the whole operator identifier string. It might be better to just use the operator id.

Pros:
- No need to handle illegal characters.
- The description string can become quiete length. Given that the dir name is also already long, we can avoid running into any cases that paths go beyond the limits of certain filessystems.

Cons:
- This string is a bit more descriptive.

What do you think?

Reply:
I think the Pros make sense to me.

If we use the operator id as the directory name, we need to extract the operator id from operator identifier text, in this case, I want to add an extract method in `OperatorSubtaskDescriptionText`.

In the other hand, I found there is a `backendUID` in `RocksIncrementalSnapshotStrategy`, How about using 'backendUID' as the directory name? If we think the `backendUID`(an UUID.randomUUID()) is too long, we can erasure all the '-' character.

What do you think?
```
operatorID.toString()         ---> d1c4061a511396721732590877b90cc6
UUID.randomUUID().toString()  ---> 3dff9490-2957-431b-a682-c7cbb995e872
after erasuring '-'           ---> 3dff94902957431ba682c7cbb995e872
```

Reply:
I would have suggested to pass the operator id into the backend instead of string operations, but you are right, using the backend UUID can also be used to fix the problem. The only downside is that we can not conclude anymore about which opertor the directory belongs to. But that might be ok.